[DRAFT] replace cross with a buildx multiarch dockerfile

CommanderStorm opened the pull request:
This is a more complex PR doing 20 things in 1.
I will need to break it apart into the elements/thoughts.

It does:
- throw away cross
- add a new dockerfile using zigbuild + buildx instead of cross + quemu + docker
- more caching for the docker build (lets see if we have to remove this. Github has an fairly small cache)
- adds attestation and sbom generation #1878 
- adds labels #1877

CommanderStorm committed:
replace the old dockerfile with an debuggable one

CommanderStorm committed:
Merge branch 'main' into new-dockerfile

nyurik reviewed (commented):
if you can pull it off.... wow :)

CommanderStorm committed:
actually add the new dockerfile

CommanderStorm committed:
Merge branch 'main' into new-dockerfile

CommanderStorm committed:
dockerfile cleanup

CommanderStorm committed:
simplify which directoryies/files are copied to an minimum subset

CommanderStorm committed:
feat(new-dockerfile): try annotating both at index and manifest level (#1879)

this is a followup to
- #1877

Apparently, github does not accept the metadata being there..
ridiculous.

this is the last PR tampering with docker metadata, I swear.

<img
src="https://media3.giphy.com/media/v1.Y2lkPWJkM2VhNTdlYXp0bmZ2aGRuZzJ6NnFyYXM4M3p3Z2FibGQ1bWt3ZG1keW9zb2VlMiZlcD12MV9naWZzX3NlYXJjaCZjdD1n/IPlurp1TLz8rrJHZwj/giphy.gif"/>

CommanderStorm committed:
merge CI with other branches

CommanderStorm committed:
apply learinigns from other pr

CommanderStorm committed:
remvoe sbom verification for the local registry

CommanderStorm committed:
Merge branch 'main' into new-dockerfile

CommanderStorm committed:
fix an merge mishap

CommanderStorm commented on .github/workflows/ci.yml:
```diff
@@ -291,13 +261,15 @@ jobs:
           labels: ${{ steps.docker_meta.outputs.labels }}
           platforms: linux/amd64,linux/arm64
+          cache-from: type=gha
+          cache-to: type=gha,mode=min
```
Looking at https://github.com/maplibre/martin/actions/caches we should not add more thigns to the cache.

This is also not increasable. I might know that guy.. small world ^^
https://github.com/orgs/community/discussions/42506#discussioncomment-11388788



```suggestion
```

CommanderStorm committed:
remove caching to gha

CommanderStorm committed:
remove benches exclusion as cargo.toml references it

CommanderStorm committed:
make sure that the rust version is specified in the build

CommanderStorm committed:
fix accidentally copying binaries to the wrong place

CommanderStorm committed:
migrate the AWS Lambda test to depend on build instead of the docker
build

CommanderStorm committed:
solved a weird dependecy cycle

CommanderStorm committed:
make sure that `test-with-svc` does not unnessesarily depend on `build`

CommanderStorm committed:
fix hidden assumption of how `actions/download-artifact` works

CommanderStorm committed:
fix another typo about how the uploading/downloading works

CommanderStorm committed:
move the dockerfile to the old location

CommanderStorm commented on .github/workflows/ci.yml:
```diff
@@ -368,17 +338,18 @@ jobs:
       - run: tests/test-aws-lambda.sh
         env:
-          MARTIN_BIN: x86_64-unknown-linux-musl/martin
+          MARTIN_BIN: target_release/x86_64-unknown-linux-gnu/martin
```
I now get why this is done this way. AWS does seem to like their musl build.

@nyurik
could you help me with why this is a different pipeline step?
I have a bit of trouble to understand why this is not just in the `docker-build-test`-step.
All dependencys already exist there => why upload and download here?

(don't look at the rest of the code, it is a WIP and pre-scrubbing)

CommanderStorm committed:
upload the musl binary again for the lambda test

CommanderStorm committed:
Save musl build artifact for Lambda test

Add step to save musl build from Docker image as artifact for later use in
Lambda testing, avoiding duplicate builds

CommanderStorm committed:
revert the target-release naming change

CommanderStorm committed:
try a different pathing aproach
